bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v1/ClientLibrariesImpl.java

```diff
@@ -272,59 +282,62 @@ private String getInline(LibraryType libraryType) {
+            Map<String, ClientLibrary> allLibraries = htmlLibraryManager.getLibraries();
+            return this.getAllResourceTypes(resourceResolver)
+                .map(resourceResolver::getResource)
+                .flatMap(res -> ClientLibrariesImpl.getClientLibraries(res, allLibraries))
+                .map(ClientLibrary::getCategories)
+                .flatMap(Arrays::stream)
+                .filter(category -> pattern == null || pattern.matcher(category).matches())
+                .collect(Collectors.toSet());
+        } catch (LoginException e) {
+            LOG.error("Cannot login as a service user", e);
+            return Collections.emptySet();
         }
+    }
 
-        for (ClientLibrary library : libraries) {
-            for (String category : library.getCategories()) {
-                if (pattern != null) {
-                    if (pattern.matcher(category).matches()) {
-                        categories.add(category);
-                    }
-                } else {
-                    categories.add(category);
-                }
-            }
-        }
-        return categories;
+    /**
+     * Get all resource types.
+     *
+     * @param resourceResolver The resource resolver.
+     * @return Stream of all resource types under which to search for client libraries.
+     */
+    private Stream<String> getAllResourceTypes(@NotNull final ResourceResolver resourceResolver) {
+        return Stream.concat(resourceTypeSet.stream(),
+            inherited
+                ? resourceTypeSet.stream().flatMap(resourceType -> Utils.getSuperTypes(resourceType, resourceResolver).stream())
+                : Stream.empty())
+            .distinct();
     }
 
     /**
-     * Adds client libraries to the provided collection, starting from the given resource
+     * Gets a stream of client libraries, starting from the given resource
      * and diving into its descendants.
      *
      * @param resource - the given resource, which will be checked to see if it's a client library
-     * @param libraries - the provided collection of libraries to add to
+     * @param allLibraries - Map of all client libraries.
      */
-    private void addClientLibraries(Resource resource, Collection<ClientLibrary> libraries) {
-        if (resource == null) {
-            return;
-        }
-        String componentType = resource.getResourceType();
-        if (StringUtils.equals(componentType, CQ_CLIENTLIBRARY_FOLDER)) {
-            ClientLibrary library = allLibraries.get(resource.getPath());
-            if (library != null) {
-                libraries.add(library);
-            }
-        }
-        for (Resource child : resource.getChildren()) {
-            addClientLibraries(child, libraries);
-        }
+    private static Stream<ClientLibrary> getClientLibraries(@org.jetbrains.annotations.Nullable final Resource resource,
```

Comment:
Here also I would prefer keeping the former implementation and not using the Java stream, which I find harder to debug and harder to follow with a recursion.